Adding RViz tutorials (#3908)
ros2/ros2_documentation: source/Tutorials/Intermediate/RViz/Marker-Display-types/Marker-Display-types.rst

Marker: Display types
=====================

**Goal:** This tutorial explains the basic Marker types and how to use them.

**Tutorial level:** Intermediate

**Time:** 15 Minutes

.. contents:: Contents

>    :depth: 3

Reviewer: ```suggestion
   :depth: 2
```

   :local:

> Backround

Reviewer: ```suggestion
Background
```

---------
The Markers display allows programmatic addition of various primitive shapes to the 3D view by sending a

> `visualization_msgs/msg/Marker <https://github.com/ros2/common_interfaces/blob/rolling/visualization_msgs/msg/Marker.msg>`_ or

Reviewer: ```suggestion
`visualization_msgs/msg/Marker <https://github.com/ros2/common_interfaces/blob/{REPOS_FILE_BRANCH}/visualization_msgs/msg/Marker.msg>`_ or
```

> `visualization_msgs/msg/MarkerArray <https://github.com/ros2/common_interfaces/blob/rolling/visualization_msgs/msg/MarkerArray.msg>`_ message.

Reviewer: ```suggestion
`visualization_msgs/msg/MarkerArray <https://github.com/ros2/common_interfaces/blob/{REPOS_FILE_BRANCH}/visualization_msgs/msg/MarkerArray.msg>`_ message.
```

.. image:: images/marker_overview.png

The :doc:`Marker: Sending Basic Shapes <../Markers-Sending-Basic-Shapes-CPP/Markers-Sending-Basic-Shapes-CPP>` tutorial begins a series of tutorials on sending markers.

The Marker Message
------------------
1 Example Usage (C++)
^^^^^^^^^^^^^^^^^^^^^
First we need to create a publisher and set a topic:

.. code-block:: C++

    auto marker_pub = node->create_publisher<visualization_msgs::msg::Marker>("visualization_marker", 1);

> After that it is as simple as filling out a `visualization_msgs/msg/Marker <https://github.com/ros2/common_interfaces/blob/rolling/visualization_msgs/msg/Marker.msg>`_

Reviewer: ```suggestion
After that it is as simple as filling out a `visualization_msgs/msg/Marker <https://github.com/ros2/common_interfaces/blob/{REPOS_FILE_BRANCH}/visualization_msgs/msg/Marker.msg>`_
```

message and publishing it:

.. code-block:: C++

    visualization_msgs::msg::Marker marker;

    marker.header.frame_id = "/my_frame";
    marker.header.stamp = rclcpp::Clock().now();

    marker.ns = "basic_shapes";
    marker.id = 0;

    marker.type = visualization_msgs::msg::Marker::SPHERE;

    marker.action = visualization_msgs::msg::Marker::ADD;

    marker.pose.position.x = 0;
    marker.pose.position.y = 0;
    marker.pose.position.z = 0;
    marker.pose.orientation.x = 0.0;
    marker.pose.orientation.y = 0.0;
    marker.pose.orientation.z = 0.0;
    marker.pose.orientation.w = 1.0;

    marker.scale.x = 1.0;
    marker.scale.y = 1.0;
    marker.scale.z = 1.0;

    marker.color.r = 0.0f;
    marker.color.g = 1.0f;
    marker.color.b = 0.0f;
    marker.color.a = 1.0;   // Don't forget to set the alpha!

>     //only if using a MESH_RESOURCE marker type:

Reviewer: ```suggestion
    // only if using a MESH_RESOURCE marker type:
```

    marker.mesh_resource = "package://pr2_description/meshes/base_v0/base.dae";

    marker.lifetime = rclcpp::Duration::from_nanoseconds(0);

    marker_pub->publish(marker);

> There is also a `visualization_msgs/msg/MarkerArray <https://github.com/ros2/common_interfaces/blob/rolling/visualization_msgs/msg/MarkerArray.msg>`_ message, which lets you publish many markers at once.

Reviewer: ```suggestion
There is also a `visualization_msgs/msg/MarkerArray <https://github.com/ros2/common_interfaces/blob/{REPOS_FILE_BRANCH}/visualization_msgs/msg/MarkerArray.msg>`_ message, which lets you publish many markers at once.
```
